frontend/userpatch.lua

```diff
@@ -0,0 +1,121 @@
+--[[--
+Allows applying developer patches while running KOReader.
+
+The contents in `koreader/userpatches/` are applied on calling `userpatch.applyPatches(priority)`.
+--]]--
+
+local isAndroid, android = pcall(require, "android")
+
+local userpatch = {
+    -- priorities for user patches,
+    early_once = "0",  -- to be started early on startup (once after an update)
+    early = "1",       -- to be started early on startup (always, but after an `early_once`)
+    late = "2",        -- to be started after UIManager is ready (always)
+                       -- 3-7 are reserved for later use
+    before_exit = "8", -- to be started a bit before exit before settings are saved (always)
+    on_exit = "9",     -- to be started right before exit (always)
+
+    -- the patch function itself
+    applyPatches = function(priority) end, -- to be overwritten, if the device allows it.
+}
+
+if isAndroid and android.prop.flavor == "fdroid" then
+    return userpatch -- allows to use applyPatches as a no-op on F-Droid flavor
+end
+
+local lfs = require("libs/libkoreader-lfs")
+local logger = require("logger")
+local DataStorage = require("datastorage")
+
+-- the directory KOReader is installed in (and runs from)
+local package_dir = lfs.currentdir()
+-- the directory where KOReader stores user data
+local data_dir = DataStorage:getDataDir() or package_dir
```

Comment:
getDataDir is never nil and should fallback to package dir (https://github.com/koreader/koreader/blob/master/datastorage.lua#L28) so there's no need to cover that part.